### ISSUE TITLE
Specify content of description field(and maybe others)

### ISSUE THREAD

**Issue by cburchert** [amiv-eth/amivapi#127]
We just had a discussion about what the event description field contains. Our quick result was that markdown would be a good option, as everyone could generate their own formating based on the environment in which the event is rendered.
It should be considered, that newsletter entries do not have much space, therefore big amounts of text are not necessary, at the same time it would be desireable to be able to include links and simple text formating.
In any case it should be specified in the docs and if possible validated.
Another important thing here is XSS. We need to either filter HTML tags or make sure developers know that the api does not prevent the inclusion of HTML tags in fields.

**Comment by hermannsblum** [comment]
Good point to raise that question.

I looked through the last newsletters and could not find any usecase of textformatting that goes beyond

* paragraphs
* bold highlighting
* links

Therefore, we should clearly document that these 3 should be supported by clients.

A big question is the matter of validation. I am note sure, but it doesn't look to me like it is possible to distinguish XSS from regular text accurately. ("<" and ">" may be used by users for other purposes, e.g. >.<). However, I think the API is at a critical point where it would be good to try to validate the content as all clients will rely on correct markup.  
One idea we could think of is some kind of "soft" validation where we would reject a request if we think it is XSS with a message like "we think you try to post XSS content. content should be formatted using markdown. If you are sure you did everythin right, just post it again".
Another option would be validation or restriction of content formatting on the client side.

What do you think?

**Comment by Leonidaz0r** [comment]
It does not make sence to reject it once, as an attacker can just send it
again and other people are randomly bothered. We have three possibilities:
1. We do not allow < and > for security, so even a client which fails to
escape characters is safe and justs displays bad characters.
2. The server HTML escapes bad characters. This is weird, as we do not even
assume HTML format on the client.
3. The client escapes HTML characters. In my opinion this is the logical
solution, as we do not assume HTML format of the client, so for example
rendering in latex, pdf or something similar is possible. We should make it
clear that the client has to escape characters. Most developers should
realize that anyway, as other characters have to be escaped as well, like &
or umlaute.
Am 10.05.2016 6:06 nachm. schrieb "Hermann" <notifications@github.com>:

**Comment by hermannsblum** [comment]
I don't understand anything what you are talking about.

What should an "attacker" be in this scenario?? Somebody who wants badly formatted text? 😃 I don't get why we should think about security of an unspecified client at this level.

Also, as I understand, this is really not about  escaping and displaying characters but about how to specify format. I think it is already specified that any api text comes as utf-8 string.

The real question is how we can ensure that texts are formatted in Markdown so that clients can rely on API content and don't have to double check it.

**Comment by NotSpecial** [comment]
Well I don't think thats too far fetched. An attacker would use his ability to insert html tags to create htmls elements that are displayed on the website with whatever intent.

Or if we leave malicious intent out of the game: Suppose We allow html. Someone uses this because he knows the website will render it and so his event text contains some html tags. But as a result all other (not browser based) tools display horrible html tags.

I support the idea to support markdown and nothing else.
On client side there are js libraries like [showdown](https://github.com/showdownjs/showdown) to convert it to html (and even unconverted it looks ok)

And on our site there are plenty of [options](http://stackoverflow.com/questions/24856035/how-to-detect-with-python-if-the-string-contains-html-code) to detect html (which we should just reject.)

But I'm not that familiar with all function of markdown: Can we just accept everything or should we filter the markdown possibilities?

**Comment by Leonidaz0r** [comment]
I just had a look at the markdown definiton and tried to get a picture of
what we need. The first thing to note is, that one of markdown's intentions
is a quick way for a writer to write HTML. It is actually a superset of
HTML, so every HTML tag is valid markdown. Markdown style is just a shorter
way to write text. While I suppose that is not used in most cases, markdown
allows any HTML tag in its code.
A consequence is escaping. Markdown also needs to escape & and < the same
as HTML, as they would be interpretiered by the layouter(as this is a
superset of HTML!).

That means we are basically using HTML format with a preprocessor if we use
markdown. We could probably reason, that no inline HTML should be used, as
the markdown syntax allows everything we need. We could therefore write a
validator, that tries to find any html tag and rejects the input if it
finds one. That would be as easy as searching for any unescaped < or & as
those must be escaped in markdown as well.

As markdown is supposed to be compiled to HTML we could also offer an
option to compile it on the server. Like if you supply compile=true as a
parameter, then we run a markdown compiler before sending it to the
client(which we could easily do on modification to save time).

One issue we have to discuss is links. Markdown has internal links, we don
not really know how those should be interpreted. We could just disallow
them and use external links. However we could also allow internal links to
other resources. That would mean more work for clients and I am not sure
anybody would ever use them, as an editor for an event for example does not
even know the API resources.
Am 14.05.2016 00:28 schrieb "Alexander Dietmüller" <notifications@github.com

**Comment by hermannsblum** [comment]
Like I wrote before, I really only see a need to support 3 different thing:

* paragraphs
* bold highlighting
* links

Yes, markdown was initially created as a short language for blog-posts which would be rendered to HTML. However, I don't like the idea that our API provides rendered html it depends a lot on the client (app, newsletter maker, newsscreen, webpage, automatic flyer printer, ...) how the markdown is rendered. I suppose it will not always be HTML and that is fine because it does not have to be.

I would really limit it to a small subset of markdown features, document them, validate them and reject everything else.

**Comment by NotSpecial** [comment]
For reference, the markdown definition @Leonidaz0r refers to can be found [here](https://daringfireball.net/projects/markdown/).

While I agree with you, @hermannsblum, that we probably won't need any more than those features, I think we'll bite ourselves if we try to enforce this.

The more rules we add, the more work we pile on ourselves. 

If we just say: Use markdown! And maybe link to the official definition we get rid of the following problems: 

* keep our documentation up to date (I think we are all aware how bad we are at really good documentation at the moment)
* arguments about which features of markdown to implement (maybe in the future someone really wants to use quotes)
* probably simpler validation rules since we just have to search for unescaped characters and `<` like @Leonidaz0r said and with that a lot less testing

My main point is still:

Markdown is very readable even unparsed. So even if the client for some reason has no library to parse markdown (unprobable. even haskell [has one](https://hackage.haskell.org/package/cheapskate)) the text will look nice.

Limiting the features just creates a lot of work for us (instead of less work) so I think we shouldn't.

Regarding links in markdown: I don't really think we should bother. If unparsed its easy to understand that those items are pointing to each other. And if the client can parse it to something nice its a plus.

What to validate exactly remains a problem though. There are just so many different kinds of text formattings that we really have no chance of filtering them all.

But I think html tags might be the only thing we really have to worry about since most of the content will be handled online and someone might be tempted to use html formatting.

In conclusion I think the easiest way will be:
* Just state that text fields use markdown format
* Use html detection like discussed [here on stack overflow](http://stackoverflow.com/questions/24856035/how-to-detect-with-python-if-the-string-contains-html-code) to reject anything that might look like a html tag

ps. I like the idea of optionally returning the text fields as html. I don't really think this will cause any problems since the default output will be unparsed markdown which is fine for every client. But specifing a header like `Format: html`or whatever could return a html formatted version. I think this would be nice since most client will be web based and this way you don't have to spend time to correctly use js to get pretty html output

**Comment by NotSpecial** [comment]
Currently implementing this, I have assigned myself.
